File: dart/day8_part1_2016.dart
import 'dart:io';

void main() {
  List<List<bool>> screen = List.generate(6, (_) => List.filled(50, false));

  List<String> instructions = File('input.txt').readAsLinesSync();

  for (String instruction in instructions) {
    if (instruction.startsWith('rect')) {
      List<int> dimensions = instruction.substring(5).split('x').map(int.parse).toList();
      for (int i = 0; i < dimensions[1]; i++) {
        for (int j = 0; j < dimensions[0]; j++) {
          screen[i][j] = true;
        }
      }
    } else if (instruction.startsWith('rotate row')) {
      List<String> parts = instruction.substring(13).split(' by ');
      int row = int.parse(parts[0]);
      int amount = int.parse(parts[1]);
      List<bool> newRow = List.generate(50, (_) => false);
      for (int i = 0; i < 50; i++) {
        newRow[(i + amount) % 50] = screen[row][i];
      }
      screen[row] = newRow;
    } else if (instruction.startsWith('rotate column')) {
      List<String> parts = instruction.substring(16).split(' by ');
      int col = int.parse(parts[0]);
      int amount = int.parse(parts[1]);
      List<bool> newCol = List.generate(6, (_) => false);
      for (int i = 0; i < 6; i++) {
        newCol[(i + amount) % 6] = screen[i][col];
      }
      for (int i = 0; i < 6; i++) {
        screen[i][col] = newCol[i];
      }
    }
  }

  int litPixels = 0;
  for (int i = 0; i < 6; i++) {
    for (int j = 0; j < 50; j++) {
      if (screen[i][j]) {
        litPixels++;
      }
    }
  }

  print(litPixels);
}

File: kotlin/day8_part1_2016.kt
import java.io.File

fun main() {
    val input = File("input.txt").readLines()

    val screen = Array(6) { BooleanArray(50) }

    input.forEach {
        val parts = it.split(" ")
        when {
            parts[0] == "rect" -> {
                val (a, b) = parts[1].split("x").map { it.toInt() }
                for (i in 0 until b) {
                    for (j in 0 until a) {
                        screen[i][j] = true
                    }
                }
            }
            parts[0] == "rotate" && parts[1] == "row" -> {
                val row = parts[2].split("=")[1].toInt()
                val shift = parts[4].toInt()
                val newRow = BooleanArray(50)
                for (i in 0 until 50) {
                    newRow[(i + shift) % 50] = screen[row][i]
                }
                screen[row] = newRow
            }
            parts[0] == "rotate" && parts[1] == "column" -> {
                val col = parts[2].split("=")[1].toInt()
                val shift = parts[4].toInt()
                val newCol = BooleanArray(6)
                for (i in 0 until 6) {
                    newCol[(i + shift) % 6] = screen[i][col]
                }
                for (i in 0 until 6) {
                    screen[i][col] = newCol[i]
                }
            }
        }
    }

    var litPixels = 0
    for (i in 0 until 6) {
        for (j in 0 until 50) {
            if (screen[i][j]) {
                litPixels++
            }
        }
    }

    println(litPixels)
}